contracts/staking/contracts/src/stake/MixinTimelockedStake.sol

```diff
@@ -0,0 +1,149 @@
+    /// |             |        0        |     0     |    0     |    0    |       0        |          0              |
+    /// |   lock(5)   |        1        |     1     |    5     |    0    |       0        |          0              |
+    /// |             |        2        |     1     |    5     |    0    |       0        |          0              |
+    /// |   lock(10)  |        2        |     2     |    15    |    5    |       0        |          0              |
+    /// |             |        3        |     2     |    15    |    5    |       0        |          5              |
+    /// |   lock(15)  |        3        |     3     |    30    |    15   |       5        |          5              |
+    /// |             |        4        |     3     |    30    |    15   |       5        |          15             |
+    /// |             |        5        |     3     |    30    |    15   |       5        |          30             |
+    /// |   lock(0)   |        5        |     5     |    30    |    30   |       30       |          30             |
+    /// |   lock(20)  |        6        |     6     |    50    |    30   |       30       |          30             |
+    /// |  unlock(30) |        6        |     6     |    20    |    0    |       0        |          0              |
+    /// |             |        7        |     6     |    20    |    0    |       0        |          0              |
+    /// |             |        8        |     6     |    20    |    0    |       0        |          20             |
+    /// -------------------------------------------------------------------------------------------------------------
+    
+    using LibSafeMath for uint256;
+
+    /// @dev Forces the timelock data structure to sync to state.
+    /// This is not necessary but may optimize some subsequent calls.
+    /// @param owner of Stake.
+    function forceTimelockSync(address owner)
+        external
+    {
+        _syncTimelockedStake(owner);
+    }
+
+    /// @dev Timelocks Stake
+    /// This moves state into the Deactivated & Timelocked state.
+    /// @param owner of Stake.
+    /// @param amount of Stake to timelock.
+    function _timelockStake(address owner, uint256 amount)
+        internal
+    {
+        (IStructs.Timelock memory ownerTimelock,) = _getSynchronizedTimelock(owner);
+        uint96 downcastAmount = amount._downcastToUint96();
+        ownerTimelock.total += downcastAmount;
+        timelockedStakeByOwner[owner] = ownerTimelock;
+    }
+
+    /// @dev Updates storage to reflect the most up-to-date timelock data structure for a given owner.
+    /// @param owner of Stake.
+    function _syncTimelockedStake(address owner)
+        internal
+    {
+        (IStructs.Timelock memory ownerTimelock, bool isOutOfSync) = _getSynchronizedTimelock(owner);
+        if (!isOutOfSync) {
+            return;
+        }
+        timelockedStakeByOwner[owner] = ownerTimelock;
```

Comment:
I think this should just be a single `sstore` since the struct occupies a single word. We do have another task to look into optimizations here. I'll add a note so we don't forget.